### ISSUE TITLE
Feature Request - Hardware keys support

### ISSUE THREAD

**Issue by scrome1337** [gcobb321/icloud3#357]
Hi, i'm getting this repair message everytime i restart my HA instance telling me to update my login information.
![CleanShot 2024-07-13 at 11 41 39@2x](https://github.com/user-attachments/assets/1225e276-fef8-405e-8c58-683e2a6efe96)

However, it doesn't matter what i choose there my icloud3 integration is synced and works like a charm, but this repair message telling me that there is a error is a bit annoying. 

Is there a way to only block this specific error message?

**Comment by gcobb321** [comment]
The iCloud account probably needs to be reauthenticatrd to clear that message. Select ‘Request a new code’. Apple will send the 6-digit code to your iPhone, enter it, submit and that should take care of it.

**Comment by scrome1337** [comment]
I would love to that, probably requested that code 100 times now.
All apple does in this situation is sending a information to my phone/mac that a new device has been login and i can click on "Ok" or "This wasn't me". So i've never ever recieved a code from clicking on "Request a new code"...

I mean the device tracking with icloud3 works fine... it's just this message which keeps coming up since i have no code to type in

**Comment by gcobb321** [comment]
Are you still having this issue with v3.0.5.5? Or should it be Closed?

**Comment by scrome1337** [comment]
just updated to v3.0.5.6, getting the same error... still can't request the code


![image](https://github.com/user-attachments/assets/cd507918-3755-4b16-88f6-cabae4cc1ea1)

**Comment by gcobb321** [comment]
Apple is still reporting that a 2fa code is needed and it is getting picked up by iCloud3.

Are you using a hardware key?

**Comment by scrome1337** [comment]
I am using a hardware key, yes.

**Comment by gcobb321** [comment]
Ahh. iCloud3 does not support hardware keys yet. I believe the official HA iCloud app does and I will be cloning that code on the v3.1 update if that is true. I'll reopen and let you know when I am testing that function and give you a preview to see if it works for you.

**Comment by scrome1337** [comment]
ty! happy to hear it

**Comment by scrome1337** [comment]
![CleanShot 2024-11-10 at 18 29 04@2x](https://github.com/user-attachments/assets/868b459c-5ecb-4680-828b-b1911661df26)


After updating to the newest version of icloud3 i keep getting the same error but this time i can't dismiss it.... Are you still planing to support hardware keys?

**Comment by gcobb321** [comment]
I have bought 3 Yubi keys and have added d topt fields to the Apple account config parameters for them. Then Apple changed the password login protocol to use Secure Remote Passwords and that blew everything up. I rushed out a v3.1 release to fix this issue since no one could login anymore. There were also lots of code changes to support multiple Apple accounts that now seems to be settling down. 

When I get back from a 3-week trip, I will be back on it.

**Comment by scrome1337** [comment]
Hi, maybe you want to take a look at this again, thank you.

**Comment by gcobb321** [comment]
I do now that things in iCloud3 are settling down and most of the issues I know about have been dealt with. I’ll send you the code when it’s ready to help check it out.

**Comment by MarcPot** [comment]
If you need more beta testers, I’m also using hardware keys.

**Comment by gcobb321** [comment]
I did a deep dive into this a couple off weeks ago and got nowhere. Apple does not provide a hook to use this but I’ll get back into it when I can. Decided the docs were severely out of date and I was spending a bunch of time answering the same questions. Then a bug or two cropped up, etc. I’ll keep you posted.

**Comment by gcobb321** [comment]
@MarcPot  @scrome1337 
I wanted to let yu know that there is some activity on supporting FIDO2 Hardware Keys on the primary PyiCloud site. PyiCloud is the code for supporting Apple account data access. I am following thae progress and will dive into it when it is ready and I have time.

Gary

**Comment by gcobb321** [comment]
@MarcPot @scrome1337 @dextorer @erkin-io @thenoid

I just uploaded v3.6 to the Development Repo (icloud3_v3) that supports Security Keys. If all goes well, I will upload it to the Regular repo later this week. Find more info [here](https://github.com/gcobb321/icloud3_v3/issues/506)

I would appreciate it if you would review the docs and try it out. Let me know how it went. 

It should be available on HACS as an update to iCloud3 Development Version . You may have to do an _Update Information_ to refresh HACS.

**Comment by dextorer** [comment]
> I just uploaded v3.6 to the Development Repo (icloud3_v3) that supports Security Keys. If all goes well, I will upload it to the Regular repo later this week. Find more info [here](https://github.com/gcobb321/icloud3_v3/issues/506)
> 
> I would appreciate it if you would review the docs and try it out. Let me know how it went.
> 
> It should be available on HACS as an update to iCloud3 Development Version . You may have to do an _Update Information_ to refresh HACS.

Just tested, and it worked flawlessly 🥳 Thank you so much for the hard work Gary, I really appreciate it!

**Comment by gcobb321** [comment]
@dextorer 
Excellent. Thanks for trying it out

**Comment by dvcrn** [comment]
I'm trying this out in the latest dev version - how did you get it working? Do you have to insert the actual security key into the machine running HASS for this to work?

**Comment by gcobb321** [comment]
Yes. See The iCloud3 docs [here](https://gcobb321.github.io/icloud3_v3_docs/#/chapters/3_apple-acct-auth) for info about the authentication methods.
